# 自分で実装
ソルバを使わずに自力で実装．茨の道．  
* 参考文献
<a name="1">小郷寛，美多勉，"システム制御理論入門"，実教出版株式会社，1979年</a>

In [72]:
using Plots
using LinearAlgebra

リカッチ代数方程式
$$
PA+A^{\top}P-PBR^{-1}B^{\top}P+Q=0
$$
の正定解$P$を求める．  

In [73]:
# テスト用のシステムを定義
test = (
    A = [
        1.1 2
        -0.3 -1
    ],
    B = [
        1 2
        0.847 3
    ],
    C = [
        1. 0.
        0. 1.
    ],
    Q = diagm([10.0, 10.0]),
    R = diagm([1., 1.])
);

ソルバを使って求めたリカッティ方程式の解

In [74]:
using MatrixEquations
P_solver, _, _ = arec(test.A, test.B, test.R, test.Q, zero(test.B))
P_solver

2×2 Matrix{Float64}:
  3.34568  -1.07266
 -1.07266   1.30235

# リカッチの非線形行列微分方程式を解いて求める
[教科書](#1)のp.160にある方法．力技．こちらは簡単．  

In [75]:
"""離散時間リカッティ方程式を解く"""
function solve_dare(A, B, Q, R, step=0.001)
    P = zero(Q)  # 初期値
    P_old = zero(Q)

    while true
        copy!(P_old, P)
        P = P .+ (P*A .+ A'*P .- P*B*inv(R)*B'*P .+ Q) .* step
        if abs.(P_old .- P) |> maximum < 1e-12
            break
        end
    end

    P
end

P = solve_dare(test.A, test.B, test.Q, test.R)
P

2×2 Matrix{Float64}:
  3.34568  -1.07266
 -1.07266   1.30235

ソルバを使って求めた解と比較

In [76]:
P - P_solver

2×2 Matrix{Float64}:
 -1.81815e-10   7.89497e-11
  7.89497e-11  -3.4283e-11

ほぼ合ってる．  

****
## 有本・ポッターの方法
[教科書](#1)のp.160にある方法．  
まずはハミルトン行列$H$をつくる．  
$$
H = \begin{bmatrix}
A & -BR^{-1}B^{\top}\\
-Q & -A^{\top}
\end{bmatrix}
$$

In [77]:
A, B, C, Q, R = test
ℋ = [
    A -B*inv(R)*B'
    -Q -A'
]  #  ハミルトン行列
ℋ

4×4 Matrix{Float64}:
   1.1    2.0  -5.0    -6.847
  -0.3   -1.0  -6.847  -9.71741
 -10.0   -0.0  -1.1     0.3
  -0.0  -10.0  -2.0     1.0

ハミルトン行列の固有値，固有ベクトルを求める．  

In [78]:
λ = eigvals(ℋ)  # ハミルトン行列の固有値
ω = eigvecs(ℋ)  # ハミルトン行列の固有ベクトル

scatter(real(λ), imag(λ), xlabel="Re", ylabel="Im", label="eigvals")
savefig("ReIm.png")

![1](ReIm.png)

In [79]:
λ  # 固有値

4-element Vector{Float64}:
 -11.862446758953242
  -2.732479989130699
   2.732479989130694
  11.862446758953245

In [80]:
ω  # 固有ベクトル

4×4 Matrix{Float64}:
 0.304554  -0.100982   0.318186   -0.574726
 0.701358   0.360764   0.0677325  -0.540943
 0.266623  -0.724832  -0.789496    0.452973
 0.586733   0.578162   0.520448    0.414592

最適極は左半面にあるもの全てである．
今回は-11.86...と-2.73...の２つが最適極となる．  
$i$番目の最適極の固有ベクトル$\omega_i$を2つに分割する．  
$$
\omega_i = \begin{bmatrix}
u_i\\
v_i
\end{bmatrix}
$$

In [81]:
# -11.16...
u1 = ω[1:2, 1]
v1 = ω[3:4, 1]

# -2.73...
u2 = ω[1:2, 2]
v2 = ω[3:4, 2];

リカッチ解は次式となる．  
$$
P = [v_1, v_2, ...,v_n][u_1, u_2, ...,u_n]^{-1}
$$

In [82]:
P_ap = [v1 v2] / [u1 u2]
P_ap

2×2 Matrix{Float64}:
  3.34568  -1.07266
 -1.07266   1.30235

他の方法で求めたものと比較する．  

In [83]:
P_solver - P_ap

2×2 Matrix{Float64}:
 -4.44089e-16  1.11022e-15
 -1.33227e-15  0.0

以上を一つの関数にまとめたもの⇓

In [84]:
"""有本・ポッター"""
function arimoto_potter(A, B, Q, R)
    
    n = size(A)[1]
    
    ℋ = [
        A -B*inv(R)*B'
        -Q -A'
    ]  #  ハミルトン行列

    λ = eigvals(ℋ)  # ハミルトン行列の固有値
    ω = eigvecs(ℋ)  # ハミルトン行列の固有ベクトル

    U = Matrix{Complex{Float64}}(undef, n, n)
    V = Matrix{Complex{Float64}}(undef, n, n)

    i = 1
    for (l, w) in zip(λ, ω)
        if real(l) < 0
            U[:, i] = ω[1:n, i]
            V[:, i] = ω[n+1:end, i]
            i += 1
        end
    end

    (V * inv(U)) |> real
end

P_arimoto = arimoto_potter(test.A, test.B, test.Q, test.R)
P_arimoto

2×2 Matrix{Float64}:
  3.34568  -1.07266
 -1.07266   1.30235

****
## 速度比較
微分方程式を解く方法のと有本・ポッター法を比較してみる．  

In [85]:
using BenchmarkTools

function bench(N)
    A = rand(N, N)
    B = rand(N, N)
    Q = diagm(rand(N))
    R = diagm(rand(N))
    A, B, Q, R
end


bench (generic function with 1 method)

N = 3

In [86]:
A, B, Q, R = bench(3)
@benchmark solve_dare(A, B, Q, R)

BenchmarkTools.Trial: 14 samples with 1 evaluation.
 Range (min … max):  331.803 ms … 428.179 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     356.768 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   362.802 ms ±  29.670 ms  ┊ GC (mean ± σ):  0.47% ± 0.78%

  ▁▁ ▁  ▁ ▁▁    ▁ ▁▁▁   ▁                       █             ▁  
  ██▁█▁▁█▁██▁▁▁▁█▁███▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁█ ▁
  332 ms           Histogram: frequency by time          428 ms <

 Memory estimate: 34.56 MiB, allocs estimate: 226478.

In [87]:
@benchmark arimoto_potter(A, B, Q, R)

BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  33.200 μs …  5.763 ms  ┊ GC (min … max): 0.00% … 99.02%
 Time  (median):     43.500 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   47.509 μs ± 58.939 μs  ┊ GC (mean ± σ):  1.20% ±  0.99%

     ▄█▇▇█▆▆▅▄▃▃▂▂▁         ▁                                 ▂
  ▃▁▁█████████████████▅▇▇█████████████▇▇▆▅▆▆▆▅▅▆▆▅▃▅▅▆▅▆▅▄▇▇▆ █
  33.2 μs      Histogram: log(frequency) by time       120 μs <

 Memory estimate: 18.81 KiB, allocs estimate: 97.

有本・ポッターの方が高速．  
N = 50

In [88]:
A, B, Q, R = bench(50)
@benchmark solve_dare(A, B, Q, R)

BenchmarkTools.Trial: 4 samples with 1 evaluation.
 Range (min … max):  1.615 s …   1.707 s  ┊ GC (min … max):  8.64% … 10.86%
 Time  (median):     1.659 s              ┊ GC (median):    10.80%
 Time  (mean ± σ):   1.660 s ± 38.118 ms  ┊ GC (mean ± σ):  10.47% ±  1.35%

  █                     █        █                        █  
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ ▁
  1.61 s         Histogram: frequency by time        1.71 s <

 Memory estimate: 1.58 GiB, allocs estimate: 168880.

In [89]:
@benchmark arimoto_potter(A, B, Q, R)

BenchmarkTools.Trial: 267 samples with 1 evaluation.
 Range (min … max):  14.463 ms … 30.340 ms  ┊ GC (min … max): 0.00% … 22.17%
 Time  (median):     18.828 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   18.770 ms ±  1.837 ms  ┊ GC (mean ± σ):  0.29% ±  2.28%

            ▁        ▁ █ ▄▂   ▅▆▃▂▅▆▃▅▄▃        ▁              
  ▃▁▁▄▃▁▄▁▅▇█▃▄▄▇▆██▅█▇█▇██▇▆███████████▇▆▆▇█▄▄▄██▁▄▁▃▁▁▁▁▃▁▃ ▄
  14.5 ms         Histogram: frequency by time        23.1 ms <

 Memory estimate: 1.05 MiB, allocs estimate: 874.